examples/platforms/utils/link_metrics.cpp

```diff
@@ -0,0 +1,238 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "link_metrics.h"
+
+#include <openthread/link_metrics.h>
+
+#include "common/clearable.hpp"
+#include "common/linked_list.hpp"
+#include "common/pool.hpp"
+#include "thread/link_quality.hpp"
+
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+
+using namespace ot;
+
+static int8_t sNoiseFloor; ///< The noise floor that would be used in Link Metrics. It should be set to the platform's
+                           ///< noise floor (measured noise floor, receiver sensitivity or a constant).
+
+class LinkMetricsDataInfo : public LinkedListEntry<LinkMetricsDataInfo>, public Clearable<LinkMetricsDataInfo>
+{
+    friend class LinkedList<LinkMetricsDataInfo>;
+    friend class LinkedListEntry<LinkMetricsDataInfo>;
+
+public:
+    /**
+     * Construtor.
+     *
+     */
+    LinkMetricsDataInfo(void) { Clear(); };
+
+    /**
+     * Set the information for this object.
+     *
+     * @param[in]  aLinkMetrics     Flags specifying what metrics to query.
+     * @param[in]  aShortAddress    Short Address of the Probing Initiator tracked by this object.
+     * @param[in]  aExtAddress      A reference to the Extended Address of the Probing Initiator tracked by this
+     *                              object.
+     *
+     */
+    void Set(const otLinkMetrics aLinkMetrics, const otShortAddress aShortAddress, const otExtAddress &aExtAddress)
+    {
+        mLinkMetrics  = aLinkMetrics;
+        mShortAddress = aShortAddress;
+        memcpy(mExtAddress.m8, aExtAddress.m8, sizeof(aExtAddress));
+    }
+
+    /**
+     * This method gets Link Metrics data stored in this object.
+     *
+     * TODO: Currently the order of Link Metircs data is fixed. Will update it to follow the order specified in TLV.
+     *
+     * @param[in]   aLqi     LQI value of the acknowledeged frame.
+     * @param[in]   aRssi    RSSI value of the acknowledged frame.
+     * @param[out]  aData    A pointer to the output buffer. @p aData MUST NOT be `nullptr`. The buffer should have 2
+     *                       bytes at least (Per spec 4.11.3.4.4.6). Otherwise the behavior would be undefined.
```

Comment:
```suggestion
     * @param[out]  aData    A pointer to the output buffer. @p aData MUST NOT be `nullptr`. The buffer must have
     *                      at least 2 bytes (per spec 4.11.3.4.4.6). Otherwise the behavior would be undefined.
```